### DISCUSSION TITLE
6/9 Discord Meeting

### DISCUSSION THREAD

**Discussion by nibanks** [microsoft/msquic#2810] [Show and tell]
> **Note**: I plan to start creating a discussion item before every Discord meeting to put together an agenda (and allow others to add to it), and then attach any notes from the discussion after the meeting.

## Agenda

- Discuss Active PRs:
  - https://github.com/microsoft/msquic/pull/2071
  - https://github.com/microsoft/msquic/pull/2685
  - https://github.com/microsoft/msquic/pull/2782
  - https://github.com/microsoft/msquic/pull/2785
  - The various Get/Set Param PRs by Daiki
- Discuss In-Progress Work:
  - https://github.com/microsoft/msquic/issues/2573
  - https://github.com/microsoft/msquic/issues/2750
  - https://github.com/microsoft/msquic/issues/1172
- Discuss Important Uncommitted Work:
  - https://github.com/microsoft/msquic/issues/1249
  - https://github.com/microsoft/msquic/issues/2418
  - https://github.com/microsoft/msquic/issues/2727
- Discuss Other Issues:
  - https://github.com/microsoft/msquic/issues/2800
  - https://github.com/microsoft/msquic/issues/2794
- Discuss Ongoing LoLa Perf Work

**Comment by dahavey** [comment]
Meeting Notes:
•	Adding API changes
•	Peer Certification
o	Serialized cert store - need to use CAPI
o	Peer Certificate CAPI Format
•	What about portable certificates?
o	Maybe: Fast Schannel(?)
o	Call it InProc Peer Certificate - maps to the Schannel PnProc flag. It maps to the Schannel InProc.
o	Need more tests so that we can make sure that we can execute the code and it doesn't crash
•	Daiki - 5 PRs out, please start getting them merged. As they are ready flag them for review.
•	Wizman/Tony - busy making slides, can make some time this month for the things we want to do.
•	BBR - Goal is to get the PR merged by the end of the month
o	Wizman, yes will work with Matt.
•	…Lost connection…
o	Just started to go through the BBRv2 code
•	It is double the size of the code - much more complex.
o	This will take some time
o	Must be labeled as preview
•	Opt-in only
o	BBR for private networks only.
•	Matt and Wizman to work together to get this checked in by EoM
o	Might need more time.
o	Check in what you have now.
•	It looks good, but needs some documentation.
	More comments
•	PRs:
o	 Advanced client cert
•	Need to test revocation flags.
•	Everything else works.
•	Close with comment - We have partial validation from HTTP team.
o	Improve Schannel Certificate retrieval speed
•	Need to merge this PR
o	More comprehensive cert testing
•	Untrusted certificate tests added(?)
	Please go through PR, check whatever boxes that can be checked. Add more if needed
o	Support Dynamic Certificate
•	Not sure if this is possible in Schannel or OpenSSL
•	Either the TLS libs support it or we would have to reprocess it.
	What happens if the cert context is different?
•	Lots of complexity
•	Ask Thomas or Mathew to try to get resolution on what is possible
•	Support callback negotiation between matching ALPNs
o	Uncertain of priority right now.
•	We are understaffed and don't want to commit without .NET buy in.
o	ALPN selection would have to be done in the handshake
o	Will this block .NET 7 from shipping?
•	MsQuicStreamSend
o	Want guarantee of serialization
•	Can we change things around to give them the serialization
•	Or can we just give them an API to query?
o	Will they consume this if we build it?
o	They call send while our thread is aborting stuff and they get the failed send call before the abort notification.
o	The send call would have to block or succeed until the abort call
•	The cleanup would have to happen after this. Right now it happens before.
•	Improving the dashboard
o	Fixed bug with shared exec context Linux
•	Wasn't really using shared exec context
o	Latency for shared exec context on Linux is really low. Better than Windows Kernel
•	Not better than XDP
•	Had meeting with UC Louvain
o	They are interested in DPDK, XDP, look into an interop runner for perf.
o	This would be very interesting. MsQuic is the fastest.
•	Are we the fastest by a lot? By a little?
•	Meeting with some academics out of Georgia Tech
o	Will see if they want to publish a paper
•	Would like to get Linux perf on par with Windows Kernel perf.
o	Linux perf is terribly bi-modal.
•	The shared EC flattens out the bi-modality
o	It might be cross-NUMAing
•	If it goes cross-NUMA then the perf is low. Otherwise the perf is good.
•	Interested in Linux, but don't have a lot of scenarios
•	MacOS is missing all of the APIs needed to be fast.
